File: apps/ios/Sources/Telemetry.swift
import Foundation
#if canImport(Sentry)
import Sentry
#endif

enum Telemetry {
    static func configure() {
#if canImport(Sentry)
        // Never report from a test run. The unit tests host the real app
        // and deliberately exercise failure paths (e.g. the login sheet's
        // "CLI not installed on broker host" fixture) — with a live DSN
        // those shipped to production Sentry as ERRORs on every CI run
        // (314 events of pure noise under release 0.0.1+13). CI no
        // longer passes a DSN to the test step; this guard is the
        // defense in depth.
        if ProcessInfo.processInfo.environment["XCTestConfigurationFilePath"] != nil
            || NSClassFromString("XCTestCase") != nil {
            return
        }
        let dsn = sentryDSN
        guard !dsn.isEmpty else { return }
        SentrySDK.start { options in
            options.dsn = dsn
            options.environment = "ios"
            options.enableMetricKit = true

            // --- Privacy: never attach PII. sendDefaultPII = false (the SDK
            // default) means Sentry does NOT collect the device IP address or
            // other personally-identifying server-side data. Set explicitly so
            // a future SDK default-flip can't silently start capturing it.
            options.sendDefaultPii = false

            // --- Session Replay, buffer mode (privacy-masked) ---------------
            // We want a short visual replay of the seconds BEFORE a crash/error
            // to diagnose on-device UI/layout bugs we can't reproduce on the
            // dev box — but NEVER an always-on screen recording, and never any
            // readable content.
            //
            //   sessionSampleRate = 0  → no always-on full-session recording.
            //   onErrorSampleRate = 1  → keep a rolling in-memory buffer and
            //                            attach it only when an event is
            //                            captured (crash/error). Nothing is
            //                            uploaded on a normal, error-free run.
            options.sessionReplay.sessionSampleRate = 0.0
            options.sessionReplay.onErrorSampleRate = 1.0
            // Mask EVERYTHING. Both default to true in 9.16.x; we pin them so a
            // default change can never un-mask. maskAllText redacts every text
            // node (chat messages, terminal output, prompts, tokens) to a solid
            // block; maskAllImages redacts every image. The replay that reaches
            // Sentry is redacted rectangles + layout only — no readable chat,
            // terminal commands, file contents, or secrets. This is the hard
            // requirement: the operator must never see user content.
            options.sessionReplay.maskAllText = true
            options.sessionReplay.maskAllImages = true
            // Replay network-body capture is opt-in only (it requires
            // options.experimental.enableReplayNetworkDetailsCapturing = true
            // PLUS a non-empty allow list). We set NEITHER, so request/response
            // bodies — which on our WebSocket carry code, terminal commands and
            // auth tokens — are never recorded. Left off deliberately.

            // --- Performance tracing + CPU profiling ------------------------
            // The point is to diagnose the on-device problems we CAN'T see from
            // the dev box: long-chat CPU burn / overheat / battery drain, slow
            // & frozen frames, slow app starts, and main-thread hangs. Auto
            // performance instrumentation (app start, slow/frozen frames,
            // UIViewController time) is on by default; we just opt into
            // sampling. 0.2 keeps overhead and quota modest while still giving a
            // representative picture.
            options.tracesSampleRate = 0.2
            // Continuous "UI Profiling" tied to the trace lifecycle: while a
            // sampled trace is active the profiler samples the call stacks, so
            // we get flame graphs of what's actually pinning the CPU during a
            // heavy chat. sessionSampleRate is RELATIVE to tracesSampleRate
            // (the legacy profilesSampleRate was removed in 9.0). Profiles are
            // stack samples (symbol names + timing) — never variable values or
            // buffer contents, so no user content leaks.
            options.configureProfiling = {
                $0.lifecycle = .trace
                $0.sessionSampleRate = 1.0
            }
            // App Hang Tracking V2 is the default (and only) implementation in
            // 9.x — the enableAppHangTrackingV2 toggle was removed in 9.0. Pin
            // the confirm-on flag + a 2s threshold so a frozen main thread
            // (the long-chat "spinner stuck" reports) surfaces as an event.
            options.enableAppHangTracking = true
            options.appHangTimeoutInterval = 2.0

            // --- Privacy: no network span/URL capture -----------------------
            // Tracing would otherwise swizzle URLSession and attach each
            // request's URL to a span. Our OAuth token exchange + API calls put
            // sensitive material in URLs/endpoints, and the WS connect URL
            // carries a `token=` secret — none of that must reach Sentry. Turn
            // the network span instrumentation OFF; we keep the perf signal we
            // actually want (CPU/frames/hangs) without any URL capture.
            options.enableNetworkTracking = false
            options.enableNetworkBreadcrumbs = false
        }
#endif
    }

    static func capture(error: Error, message: String, tags: [String: String] = [:], extras: [String: String] = [:]) {
#if canImport(Sentry)
        guard !sentryDSN.isEmpty else { return }
        SentrySDK.configureScope { scope in
            tags.forEach { scope.setTag(value: $0.value, key: $0.key) }
            extras.forEach { scope.setExtra(value: $0.value, key: $0.key) }
            scope.setLevel(.error)
        }
        SentrySDK.capture(message: message)
        SentrySDK.capture(error: error)
#else
        _ = (error, message, tags, extras)
#endif
    }

    /// Leave a lightweight breadcrumb in the trail attached to the NEXT
    /// event (crash/error). Unlike `debug`, a breadcrumb is NOT a full
    /// Sentry event — it's ring-buffered by the SDK and costs nothing
    /// until something fails, so scatter these liberally across any flow
    /// or screen that can fail (see CLAUDE.md "Standing order").
    /// `category` groups crumbs (e.g. "agent_login", "session",
    /// "connect", "browser", "upload"); `data` is searchable key/values.
    static func breadcrumb(_ category: String, _ message: String, data: [String: String] = [:]) {
#if canImport(Sentry)
        guard !sentryDSN.isEmpty else { return }
        let crumb = Breadcrumb(level: .info, category: category)
        crumb.message = message
        if !data.isEmpty { crumb.data = data }
        SentrySDK.addBreadcrumb(crumb)
#else
        _ = (category, message, data)
#endif
    }

    /// Structured diagnostic telemetry meant to be READ BACK from Sentry
    /// (org `swe-kitty`, project `conduit-ios`): an INFO-level event tagged
    /// `diag=<category>` with `data` as searchable extras. Use it for runtime
    /// state that can't be reproduced on the dev box — layout / render /
    /// timing / keyboard — so the actual on-device numbers can be read
    /// remotely instead of asked for and transcribed.
    ///
    /// Standing practice: instrument new features with `Telemetry.debug` so
    /// they're always debuggable from Sentry. It is meant to be LOW VOLUME —
    /// every call is a full Sentry event, so a high-frequency caller (keyboard
    /// show/hide, terminal resize) would otherwise flood the project, burn
    /// quota, and pile main-thread work behind the SDK. Two guards below keep
    /// that safe regardless of the caller:
    ///   1. consecutive-identical events for a category are dropped (only a
    ///      *distinct state* gets through), and
    ///   2. the event is built + submitted off the main thread.
    static func debug(_ category: String, _ message: String, data: [String: String] = [:]) {
#if canImport(Sentry)
        guard !sentryDSN.isEmpty else { return }

        // Collapse repeats: skip when this category's payload is identical to
        // the last one we sent for it. `data` is captured by value here, so
        // the comparison + dispatch are safe to run off the calling thread.
        let payload = message + "\u{1}" + data.sorted { $0.key < $1.key }
            .map { "\($0.key)=\($0.value)" }
            .joined(separator: "\u{1}")
        debugDedupeLock.lock()
        let isRepeat = lastDebugPayload[category] == payload
        if !isRepeat { lastDebugPayload[category] = payload }
        debugDedupeLock.unlock()
        guard !isRepeat else { return }

        debugQueue.async {
            SentrySDK.capture(message: "[\(category)] \(message)") { scope in
                scope.setLevel(.info)
                scope.setTag(value: category, key: "diag")
                // Collapse ALL events of a diag category into a SINGLE Sentry
                // issue (otherwise each distinct message — "keyboard will show",
                // "...will hide", "composer focused" — files its own issue and
                // floods the project). One issue per category, many events.
                scope.setFingerprint(["diag", category])
                data.forEach { scope.setExtra(value: $0.value, key: $0.key) }
            }
        }
#else
        _ = (category, message, data)
#endif
    }

#if canImport(Sentry)
    /// Serial queue so diagnostic events never cost the main thread time, even
    /// for the scope-building closure. The Sentry SDK is itself thread-safe.
    private static let debugQueue = DispatchQueue(label: "sh.nikhil.conduit.telemetry.debug", qos: .utility)
    /// Last payload emitted per `diag` category, used to drop consecutive
    /// duplicates. Guarded by `debugDedupeLock` since `debug` is called from
    /// arbitrary threads (keyboard notifications, layout passes).
    private static var lastDebugPayload: [String: String] = [:]
    private static let debugDedupeLock = NSLock()
#endif

    private static var sentryDSN: String {
        let raw = Bundle.main.object(forInfoDictionaryKey: "SentryDSN") as? String ?? ""
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        // ONLY a real DSN URL enables Sentry. Anything else must disable it
        // cleanly: the unsubstituted build placeholder `$(SENTRY_DSN)`, an
        // empty secret, or a stray value like "-" (a bad/empty SENTRY_DSN_IOS
        // secret literally shipped `SentryDSN = "-"` in v0.0.76–78, which
        // passed the old `!= "$(SENTRY_DSN)"` check, reached SentrySDK.start as
        // an invalid DSN, failed SDK init, and silently dropped EVERY event —
        // iOS telemetry went dark for three releases). Validating the URL shape
        // here turns a bad secret into "Sentry off" instead of "Sentry broken".
        guard trimmed.hasPrefix("https://") || trimmed.hasPrefix("http://") else { return "" }
        return trimmed
    }
}

File: apps/android/app/src/main/kotlin/sh/nikhil/conduit/Telemetry.kt
package sh.nikhil.conduit

import android.content.Context
import io.sentry.Breadcrumb
import io.sentry.Sentry
import io.sentry.SentryLevel
import io.sentry.android.core.SentryAndroid

object Telemetry {
    fun configure(context: Context) {
        val dsn = BuildConfig.SENTRY_DSN.trim()
        if (dsn.isEmpty()) return
        SentryAndroid.init(context) { options ->
            options.dsn = dsn
            options.environment = "android"
            options.isEnableAutoSessionTracking = true
            // versionName is pinned to "0.0.1" (device bug #7), so Sentry's
            // default release would mislabel EVERY android event
            // "sh.nikhil.conduit@0.0.1+12" — we couldn't tell which build a
            // crash came from. Use the real release tag the CI build injects.
            if (BuildConfig.RELEASE_TAG != "dev") {
                options.release = "sh.nikhil.conduit@${BuildConfig.RELEASE_TAG}"
            }

            // --- Privacy: never attach PII. isSendDefaultPii = false (the SDK
            // default) keeps Sentry from collecting the device IP address or
            // other personally-identifying data. Set explicitly so a future
            // default-flip can't silently start capturing it.
            options.isSendDefaultPii = false

            // --- Session Replay, buffer mode (privacy-masked) — parity with
            // iOS (Telemetry.swift). Short visual replay of the seconds BEFORE
            // a crash/error to diagnose on-device UI bugs we can't reproduce on
            // the dev box; never an always-on recording, never readable content.
            //
            //   sessionSampleRate = 0 -> no always-on full-session recording.
            //   onErrorSampleRate = 1 -> keep a rolling buffer, attach it only
            //                           when an event is captured. Nothing is
            //                           uploaded on a normal, error-free run.
            options.sessionReplay.sessionSampleRate = 0.0
            options.sessionReplay.onErrorSampleRate = 1.0
            // Mask EVERYTHING. Both default to true; pinned so a default change
            // can never un-mask. maskAllText redacts every text node (chat,
            // terminal output, prompts, tokens); maskAllImages redacts every
            // image. The replay that reaches Sentry is redacted rectangles +
            // layout only — no readable user content. Hard requirement: the
            // operator must never see chat logs, terminal commands, or secrets.
            // Setter-only in sentry-android 8.x (no getter -> no synthesized
            // Kotlin property), and already the SDK default — pinned explicitly
            // so a default change can never un-mask.
            options.sessionReplay.setMaskAllText(true)
            options.sessionReplay.setMaskAllImages(true)
            // Replay network-body capture is opt-in (empty networkDetailAllowUrls
            // by default = nothing captured) and relies on the Sentry OkHttp
            // interceptor, which we don't install. Bodies — which carry code,
            // terminal commands and tokens — are never recorded. Left off.

            // --- Performance tracing + CPU profiling (parity with iOS) ------
            // Diagnose the on-device problems invisible from the dev box:
            // long-chat CPU burn / overheat / battery, slow & frozen frames,
            // slow app starts, ANRs. Auto activity/app-start/frames
            // instrumentation is on by default; we opt into sampling.
            // Reduced from 0.2 -> 0.1 to trim quota; still representative.
            options.tracesSampleRate = 0.1
            // Continuous "UI Profiling" tied to the trace lifecycle (GA in
            // 8.7): while a sampled trace is active the profiler samples call
            // stacks, giving flame graphs of what pins the CPU during a heavy
            // chat. profileSessionSampleRate is relative to tracesSampleRate;
            // the legacy profilesSampleRate must NOT be set alongside it.
            // Profiles are stack samples (symbols + timing), never user content.
            // Reduced from 1.0 -> 0.1 to trim profile-upload quota.
            options.profileSessionSampleRate = 0.1
            options.profileLifecycle = io.sentry.ProfileLifecycle.TRACE
            // ANR / watchdog detection (AnrV2) is on by default; pin it. A
            // frozen main thread (the long-chat "spinner stuck" reports)
            // surfaces as an ANR event.
            options.isAnrEnabled = true

            // --- Privacy: no network URL capture ----------------------------
            // Our OAuth token exchange + WS connect carry sensitive material in
            // URLs (the WS connect URL has a `token=` secret). Drop network
            // breadcrumbs so no request URL reaches Sentry. We don't install the
            // OkHttp span interceptor, and maxRequestBodySize stays NONE
            // (default), so no body or URL is ever attached.
            options.isEnableNetworkEventBreadcrumbs = false

            // --- Quota denylist: beforeSend backstop -------------------------
            // Drop useless high-frequency events BEFORE they are uploaded.
            // This is the durable quota guard: even if a caller forgets to use
            // breadcrumb instead of debug/diagnostic, denylisted events never
            // reach Sentry.
            //
            // Rules (return null = drop):
            //   1. Any INFO-level "diag" event whose category is in the
            //      HIGH_FREQUENCY_DIAG_CATEGORIES set (keyboard, layout, scroll,
            //      terminal-resize). These are pure UI churn — no value as
            //      standalone events.
            //   2. Any event whose message matches a known routine-noise pattern
            //      (connection churn, routine disconnects). These are expected
            //      lifecycle, not errors.
            // ERROR/FATAL-level events bypass the denylist entirely.
            options.beforeSend = { event, _ ->
                // Never drop real errors.
                val level = event.level
                if (level == SentryLevel.ERROR || level == SentryLevel.FATAL) return@beforeSend event

                // Drop high-frequency UI diag categories (INFO-level).
                val diagCategory = event.tags?.get("diag")
                if (diagCategory != null && diagCategory in HIGH_FREQUENCY_DIAG_CATEGORIES) {
                    return@beforeSend null
                }

                // Drop routine connection-churn messages.
                val msg = event.message?.formatted ?: event.message?.message ?: ""
                if (isRoutineNoiseMessage(msg)) return@beforeSend null

                event
            }
        }
    }

    fun capture(error: Throwable, message: String, tags: Map<String, String> = emptyMap(), extras: Map<String, String> = emptyMap()) {
        val dsn = BuildConfig.SENTRY_DSN.trim()
        if (dsn.isEmpty()) return
        Sentry.withScope { scope ->
            scope.level = SentryLevel.ERROR
            tags.forEach { (key, value) -> scope.setTag(key, value) }
            extras.forEach { (key, value) -> scope.setExtra(key, value) }
            Sentry.captureMessage(message, SentryLevel.ERROR)
            Sentry.captureException(error)
        }
    }

    /**
     * Capture a non-error diagnostic (no exception attached) — for paths
     * that fail by returning null rather than throwing (e.g. a QR image
     * that simply doesn't decode). Lands as a WARNING-level Sentry message
     * with the supplied extras so we can see *why* it failed on a device.
     */
    fun diagnostic(message: String, tags: Map<String, String> = emptyMap(), extras: Map<String, String> = emptyMap()) {
        val dsn = BuildConfig.SENTRY_DSN.trim()
        if (dsn.isEmpty()) return
        Sentry.withScope { scope ->
            scope.level = SentryLevel.WARNING
            tags.forEach { (key, value) -> scope.setTag(key, value) }
            extras.forEach { (key, value) -> scope.setExtra(key, value) }
            Sentry.captureMessage(message, SentryLevel.WARNING)
        }
    }

    /**
     * Structured diagnostic telemetry meant to be READ BACK from Sentry
     * (org `swe-kitty`, project `conduit-android`): an INFO-level event tagged
     * `diag=<category>` with `data` as searchable extras. Use it for runtime
     * state that can't be reproduced on the dev box — layout / render /
     * timing — so the on-device numbers can be read remotely instead of asked
     * for and transcribed.
     *
     * Standing practice: instrument new features with `Telemetry.debug` so
     * they're always debuggable from Sentry. It is meant to be LOW VOLUME —
     * every call is a full Sentry event, so a high-frequency caller would
     * otherwise flood the project and burn quota. THREE guards below keep that
     * safe regardless of the caller:
     *   1. consecutive-identical events for a category are dropped (only a
     *      *distinct state* gets through),
     *   2. per-category time throttle: at most 1 event per 60s per category
     *      (token-bucket style, thread-safe via ConcurrentHashMap + atomic
     *      compare), and
     *   3. the beforeSend denylist drops high-frequency categories even if
     *      these guards are bypassed.
     *
     * Callers that fire per-interaction (keyboard, layout, scroll,
     * terminal-resize) MUST use [breadcrumb] instead — those categories are
     * also in the beforeSend denylist as a backstop.
     */
    fun debug(category: String, message: String, data: Map<String, String> = emptyMap()) {
        if (BuildConfig.SENTRY_DSN.trim().isEmpty()) return

        val payload = buildString {
            append(message)
            data.entries.sortedBy { it.key }.forEach {
                append("::").append(it.key).append('=').append(it.value)
            }
        }

        val nowMs = System.currentTimeMillis()
        synchronized(debugLock) {
            val isRepeat = lastDebugPayload[category] == payload
            val lastMs = lastDebugTimestampMs[category] ?: 0L
            val tooSoon = (nowMs - lastMs) < DEBUG_THROTTLE_MS
            if (isRepeat || tooSoon) return
            lastDebugPayload[category] = payload
            lastDebugTimestampMs[category] = nowMs
        }

        Sentry.withScope { scope ->
            scope.level = SentryLevel.INFO
            scope.setTag("diag", category)
            // Collapse all events of a diag category into a SINGLE Sentry issue
            // (otherwise each distinct message files its own issue and floods
            // the project). One issue per category, many events.
            scope.fingerprint = listOf("diag", category)
            data.forEach { (key, value) -> scope.setExtra(key, value) }
            Sentry.captureMessage("[$category] $message", SentryLevel.INFO)
        }
    }

    /**
     * Leave a lightweight breadcrumb in the trail attached to the NEXT
     * event (crash/error). Unlike [debug], a breadcrumb is NOT a full
     * Sentry event — it's ring-buffered by the SDK and costs nothing until
     * something fails, so scatter these liberally across any flow or
     * screen that can fail (see CLAUDE.md "Standing order"). [category]
     * groups crumbs (e.g. "agent_login", "session", "connect", "browser").
     */
    fun breadcrumb(category: String, message: String, data: Map<String, String> = emptyMap()) {
        if (BuildConfig.SENTRY_DSN.trim().isEmpty()) return
        val crumb = Breadcrumb(message)
        crumb.category = category
        crumb.level = SentryLevel.INFO
        data.forEach { (key, value) -> crumb.setData(key, value) }
        Sentry.addBreadcrumb(crumb)
    }

    // -------------------------------------------------------------------------
    // beforeSend denylist constants (Android parity with iOS Telemetry.swift)
    // -------------------------------------------------------------------------

    /**
     * High-frequency UI diag categories that MUST NOT produce standalone Sentry
     * events. Callers in these categories should use [breadcrumb] instead;
     * [beforeSend] drops any that slip through.
     *
     * Extend this list whenever a new per-frame / per-interaction category is
     * added. Mirror of iOS [Telemetry.highFrequencyDiagCategories].
     */
    val HIGH_FREQUENCY_DIAG_CATEGORIES: Set<String> = setOf(
        "keyboard",        // per-show/hide/focus events
        "layout",          // per-layout-pass debug
        "scroll",          // scroll-position diag
        "terminal_resize", // terminal window resize
        "terminal-resize", // alternate spelling guard
        "frame",           // per-frame render diag
    )

    /**
     * Substrings whose presence in an event message indicates routine
     * operational churn — expected lifecycle, not an actionable failure.
     * Matched case-insensitively. Mirror of iOS [Telemetry.routineNoisePatterns].
     */
    val ROUTINE_NOISE_PATTERNS: List<String> = listOf(
        "disconnected from harness",  // routine WS lifecycle
        "sessionstore: code 0",       // NSError domain=SessionStore code=0
        "code 0",
        "still disconnected",
        "keyboard will hide",
        "keyboard will show",
        "composer focused",
        "composer blurred",
    )

    /** Returns true when the message matches a routine-noise pattern. */
    fun isRoutineNoiseMessage(message: String): Boolean {
        val lower = message.lowercase()
        return ROUTINE_NOISE_PATTERNS.any { lower.contains(it) }
    }

    /** Max one debug event per category per this interval (milliseconds). */
    const val DEBUG_THROTTLE_MS: Long = 60_000L

    /** Last payload emitted per `diag` category, used to drop consecutive duplicates. */
    private val lastDebugPayload = mutableMapOf<String, String>()
    /** Timestamp (ms) of the last event actually sent per category. */
    private val lastDebugTimestampMs = mutableMapOf<String, Long>()
    /** Guards both lastDebugPayload and lastDebugTimestampMs. */
    private val debugLock = Any()
}
